lib/pwauth/ldap/impl.go

```diff
@@ -0,0 +1,90 @@
+package ldap
+
+import (
+	"crypto/x509"
+	//"errors"
+	"fmt"
+	"time"
+
+	"github.com/Symantec/Dominator/lib/log"
+	"github.com/Symantec/keymaster/lib/authutil"
+)
+
+const defaultCacheDuration = time.Hour * 96
+
+func newAuthenticator(urllist []string, bindPattern []string,
+	timeoutSecs uint, rootCAs *x509.CertPool, logger log.DebugLogger) (
+	*PasswordAuthenticator, error) {
+	var authenticator PasswordAuthenticator
+	for _, stringURL := range urllist {
+		url, err := authutil.ParseLDAPURL(stringURL)
+		if err != nil {
+			return nil, err
+		}
+		authenticator.ldapURL = append(authenticator.ldapURL, url)
+	}
+	authenticator.bindPattern = bindPattern
+	authenticator.timeoutSecs = timeoutSecs
+	authenticator.rootCAs = rootCAs
+	authenticator.logger = logger
+	authenticator.expirationDuration = defaultCacheDuration
+	authenticator.cachedCredentials = make(map[string]cacheCredentialEntry)
+	return &authenticator, nil
+}
+
+func convertToBindDN(username string, bind_pattern string) string {
+	return fmt.Sprintf(bind_pattern, username)
+}
+
+func (pa *PasswordAuthenticator) passwordAuthenticate(username string,
+	password []byte) (valid bool, err error) {
+	valid = false
+	//for _, ldapUrl := range strings.Split(config.Ldap.LDAP_Target_URLs, ",") {
+	for _, u := range pa.ldapURL {
+		for _, bindPattern := range pa.bindPattern {
```

Comment:
Given the size of what's in the outer control block, it would be better to move this to a function.